src/code/z_lifemeter.c

```diff
@@ -1,17 +1,437 @@
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_lifemeter/LifeMeter_IsCritical.s")
+    interfaceCtx->lifeColorChange = 0;
+    interfaceCtx->lifeColorChangeDirection = 0;
+
+    interfaceCtx->lifeSizeChange = interfaceCtx->lifeColorChange;
+    interfaceCtx->lifeSizeChangeDirection = interfaceCtx->lifeColorChangeDirection;
+
+    interfaceCtx->heartsPrimR[0] = 255;
+    interfaceCtx->heartsPrimG[0] = 70;
+    interfaceCtx->heartsPrimB[0] = 50;
+
+    interfaceCtx->heartsEnvR[0] = 50;
+    interfaceCtx->heartsEnvG[0] = 40;
+    interfaceCtx->heartsEnvB[0] = 60;
+
+    interfaceCtx->heartsPrimR[1] = 255;
+    interfaceCtx->heartsPrimG[1] = 70;
+    interfaceCtx->heartsPrimB[1] = 50;
+
+    interfaceCtx->heartsEnvR[1] = 50;
+    interfaceCtx->heartsEnvG[1] = 40;
+    interfaceCtx->heartsEnvB[1] = 60;
+
+    sHeartsDDPrim[0][0] = sHeartsDDPrim[1][0] = 255;
+    sHeartsDDPrim[0][1] = sHeartsDDPrim[1][1] = 255;
+    sHeartsDDPrim[0][2] = sHeartsDDPrim[1][2] = 255;
+
+    sHeartsDDEnv[0][0] = sHeartsDDEnv[1][0] = 200;
+    sHeartsDDEnv[0][1] = sHeartsDDEnv[1][1] = 0;
+    sHeartsDDEnv[0][2] = sHeartsDDEnv[1][2] = 0;
+}
+
+void LifeMeter_UpdateColors(GlobalContext* globalCtx) {
+    InterfaceContext* interfaceCtx = &globalCtx->interfaceCtx;
+    f32 factorBeating = interfaceCtx->lifeColorChange * 0.1f;
+    f32 ddFactor;
+    s32 type = 0;
+    s32 ddType;
+    s16 rFactor;
+    s16 gFactor;
+    s16 bFactor;
+
+    if (interfaceCtx) {}
+
+    if (interfaceCtx->lifeColorChangeDirection != 0) {
+        interfaceCtx->lifeColorChange--;
+        if (interfaceCtx->lifeColorChange <= 0) {
+            interfaceCtx->lifeColorChange = 0;
+            interfaceCtx->lifeColorChangeDirection = 0;
+        }
+    } else {
+        interfaceCtx->lifeColorChange++;
+        if (interfaceCtx->lifeColorChange >= 10) {
+            interfaceCtx->lifeColorChange = 10;
+            interfaceCtx->lifeColorChangeDirection = 1;
+        }
+    }
+
+    ddFactor = factorBeating;
+
+    interfaceCtx->heartsPrimR[0] = 255;
+    interfaceCtx->heartsPrimG[0] = 70;
+    interfaceCtx->heartsPrimB[0] = 50;
+
+    interfaceCtx->heartsEnvR[0] = 50;
+    interfaceCtx->heartsEnvG[0] = 40;
+    interfaceCtx->heartsEnvB[0] = 60;
+
+    interfaceCtx->heartsPrimR[1] = sHeartsPrimColors[type][0];
+    interfaceCtx->heartsPrimG[1] = sHeartsPrimColors[type][1];
+    interfaceCtx->heartsPrimB[1] = sHeartsPrimColors[type][2];
+
+    interfaceCtx->heartsEnvR[1] = sHeartsEnvColors[type][0];
+    interfaceCtx->heartsEnvG[1] = sHeartsEnvColors[type][1];
+    interfaceCtx->heartsEnvB[1] = sHeartsEnvColors[type][2];
+
+    rFactor = sHeartsPrimFactors[0][0] * factorBeating;
+    gFactor = sHeartsPrimFactors[0][1] * factorBeating;
+    bFactor = sHeartsPrimFactors[0][2] * factorBeating;
+
+    interfaceCtx->beatingHeartPrim[0] = (u8)(rFactor + 255) & 0xFF;
+    interfaceCtx->beatingHeartPrim[1] = (u8)(gFactor + 70) & 0xFF;
+    interfaceCtx->beatingHeartPrim[2] = (u8)(bFactor + 50) & 0xFF;
```

Comment:
Is the ` & 0xFF` needed to match?
```suggestion
    interfaceCtx->beatingHeartPrim[0] = (u8)(rFactor + 255);
    interfaceCtx->beatingHeartPrim[1] = (u8)(gFactor + 70);
    interfaceCtx->beatingHeartPrim[2] = (u8)(bFactor + 50);
```

Reply:
yes, I couldn't get it to match without those sadly, looks a bit odd. maybe I'm missing some compiler knowledge that might make it work without those.